slam toolbox causes bugs

**Description**

Using rclnodejs with slam-toolbox creates bug, that makes impossible to generate ros messages.

When installing rclnodejs this error does not occur but my custom ros messages are missing TypeScript definitions. Applying proposed patch fixes this problem. 

- Library Version: 0.26.0
- ROS Version: iron
- Platform / OS: ubuntu 22.04.3 LTS

**Steps To Reproduce**
```
sudo apt install ros-iron-slam-toolbox
```
and in new console:
```
npx generate-ros-messages
```


**Expected Behavior**
```
Start generation of ROS2 JavaScript messages...
Generation complete.
```


**Actual Behavior**

```
Start generation of ROS2 JavaScript messages...
Caught error: Error: Cannot find module '../../generated/slam_toolbox/slam_toolbox__srv__AddSubmap_Request.js'
Require stack:
- /home/k/backend/node_modules/rclnodejs/generated/slam_toolbox/slam_toolbox__srvs__AddSubmap_Event.js
- /home/k/backend/node_modules/rclnodejs/lib/interface_loader.js
- /home/k/backend/node_modules/rclnodejs/rostsd_gen/index.js
- /home/k/backend/node_modules/rclnodejs/scripts/generate_messages.js
```


**quick fix:**
use 'patch-package' package with this file 
rclnodejs+0.26.0.patch
```
diff --git a/node_modules/rclnodejs/rostsd_gen/index.js b/node_modules/rclnodejs/rostsd_gen/index.js
index dbb9b64..6aca532 100644
--- a/node_modules/rclnodejs/rostsd_gen/index.js
+++ b/node_modules/rclnodejs/rostsd_gen/index.js
@@ -64,6 +64,7 @@ function getPkgInfos(rootDir) {
 
     for (let filename of files) {
       const typeClass = fileName2Typeclass(filename);
+      if (typeClass.type !== 'srv' && typeClass.type !== 'action' && typeClass.type !== 'msg') continue;
       if (
         !typeClass.type ||
         pkgFilters.matchesAny({
```

--- a/rosidl_gen/idl_generator.js
+++ b/rosidl_gen/idl_generator.js
@@ -92,6 +92,9 @@ async function generateServiceEventMsg(serviceInfo, dir) {
 
 async function generateServiceEventJSStruct(msgInfo, dir) {
   const spec = await parser.parseMessageFile(msgInfo.pkgName, msgInfo.filePath);
+  // Pass `msgInfo.subFolder` to the `spec`, because some .srv files of the
+  // package may not be put under srv/ folder, e.g., slam_toolbox.
+  spec.subFolder = msgInfo.subFolder;
 
   // Remove the `.msg` files generated in `generateServiceEventMsg()` to avoid
   // being found later.

--- a/rosidl_gen/templates/message.dot
+++ b/rosidl_gen/templates/message.dot
@@ -198,7 +198,7 @@ function getModulePathByType(type, messageInfo) {
 
   /* We should use '__msg__' to require "service_msgs/msg/ServiceEventInfo.msg" for service event message. */
   if (messageInfo && messageInfo.isServiceEvent && (type.type !== 'ServiceEventInfo')) {
-    return type.pkgName + '__srv__' + type.type + '.js';
+    return type.pkgName + `__${it.spec.subFolder}__` + type.type + '.js';
   }
   return type.pkgName + '__msg__' + type.type + '.js';
 }

